src/core/graphql/gql-execution-context.host.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  CallHandler,
+  ExecutionContext,
+  Injectable,
+  NestInterceptor,
+  OnModuleDestroy,
+} from '@nestjs/common';
+import { GqlExecutionContext } from '@nestjs/graphql';
+import { AsyncLocalStorage } from 'async_hooks';
+import { GraphQLResolveInfo } from 'graphql';
+import { lastValueFrom } from 'rxjs';
+import { GqlContextType } from '../../common';
+
+export interface GqlArgumentsHost {
+  getArgs: <T = any>() => T;
+  getRoot: <T = any>() => T;
+  getInfo: () => GraphQLResolveInfo;
+  getContext: () => GqlContextType;
+}
+
+/**
+ * A service holding the current execution context
+ */
+export abstract class GqlExecutionContextHost {
+  /**
+   * The current execution context
+   */
+  readonly context: GqlArgumentsHost;
+}
+
+/**
+ * This is necessary to allow global pipes to have access to request info.
+ * At least until this is resolved: https://github.com/nodejs/node/issues/43148
+ */
+@Injectable()
+export class GqlExecutionContextHostImpl
+  implements GqlExecutionContextHost, NestInterceptor, OnModuleDestroy
+{
+  als = new AsyncLocalStorage<GqlExecutionContext>();
+
+  get context() {
+    const context = this.als.getStore();
+    if (context) {
+      return context;
+    }
+    const [major] = process.versions.node.split('.').map(Number);
+    if (major === 18) {
+      throw new Error(`The execution context has not been attached yet.
+
+There's a bug with this process when using Node 18 with debugger ATTACHED (just listening is fine).
+Please downgrade to Node 16 for now if you want to debug and are encountering this.
+
+$ brew unlink node && brew install node@16 && brew link --overwrite node@16
+`);
+    }
+    throw new Error('The execution context has not been attached yet.');
+  }
+
+  intercept(context: ExecutionContext, next: CallHandler) {
+    const gqlContext = GqlExecutionContext.create(context);
+    return this.als.run(gqlContext, () => lastValueFrom(next.handle()));
```

Comment:
This doesn't work because the interceptor only spans the query/mutation method, not the fields below it.
The info that this provides also doesn't makes sense across every field either as each would have different args, resolve info, etc.

So this needs to be re-thought.

Reply:
Middleware spans the entire request, so I switched to that in conjunction with an apollo plugin. The middleware correctly scopes the async storage and holds a ref to an object which is hydrated "later" with the apollo plugin event.